### PR TITLE
bootstrap: Add doctests and unitests

### PR DESCRIPTION
**Pull request by milmazz**
This commit includes the following changes:

* Include more docstrings in classes, methods, and functions
* Add doctests, which are great for self-documenting our source code
* Add some unit tests with the `unittest` module
* Remove `WindowsError` reference on non-windows systems
* Rename some variables to be more explicit about their meaning
* Move all the attributes defined outside of `__init__`
* Add initial support for Python 3

r? @alexcrichton

### CODE REVIEW

**Comment by milmazz** [comment]
In this commit 26885c8 you'll find initial support for Python 3. I did some tests with [tox](https://tox.readthedocs.io/) including the current unit tests for Python 2.7, 3.5, and 3.6 and everything seems to be working:

```console
$ cd src/bootstrap
$ tox
GLOB sdist-make: /Development/rust/rust/src/bootstrap/setup.py
py27 create: /Development/rust/rust/src/bootstrap/.tox/py27
py27 inst: /Development/rust/rust/src/bootstrap/.tox/dist/rust-bootstrap-0.0.1.zip
py27 installed: rust-bootstrap==0.0.1
py27 runtests: PYTHONHASHSEED='3410238989'
py27 runtests: commands[0] | /Development/rust/rust/src/bootstrap/.tox/py27/bin/python bootstrap_test.py
bin_root (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bin_root ... ok
bootstrap_binary (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bootstrap_binary ... ok
cargo_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.cargo_stamp ... ok
get_mk (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_mk ... ok
get_string (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_string ... ok
get_toml (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_toml ... ok
rustc_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.rustc_stamp ... ok
format_build_time (bootstrap)
Doctest: bootstrap.format_build_time ... ok
test_stage0_data (__main__.Stage0DataTestCase)
Extract data from stage0.txt ... ok
test_invalid_file (__main__.VerifyTestCase)
Should verify that the file is invalid ... invalid checksum:
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
ok
test_valid_file (__main__.VerifyTestCase)
Check if the sha256 sum of the given file is valid ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.013s

OK
py35 create: /Development/rust/rust/src/bootstrap/.tox/py35
py35 inst: /Development/rust/rust/src/bootstrap/.tox/dist/rust-bootstrap-0.0.1.zip
py35 installed: rust-bootstrap==0.0.1
py35 runtests: PYTHONHASHSEED='3410238989'
py35 runtests: commands[0] | /Development/rust/rust/src/bootstrap/.tox/py35/bin/python bootstrap_test.py
bin_root (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bin_root ... ok
bootstrap_binary (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bootstrap_binary ... ok
cargo_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.cargo_stamp ... ok
get_mk (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_mk ... ok
get_string (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_string ... ok
get_toml (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_toml ... ok
rustc_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.rustc_stamp ... ok
format_build_time (bootstrap)
Doctest: bootstrap.format_build_time ... ok
test_stage0_data (__main__.Stage0DataTestCase)
Extract data from stage0.txt ... ok
test_invalid_file (__main__.VerifyTestCase)
Should verify that the file is invalid ... invalid checksum:
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
ok
test_valid_file (__main__.VerifyTestCase)
Check if the sha256 sum of the given file is valid ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.021s

OK
py36 create: /Development/rust/rust/src/bootstrap/.tox/py36
py36 inst: /Development/rust/rust/src/bootstrap/.tox/dist/rust-bootstrap-0.0.1.zip
py36 installed: rust-bootstrap==0.0.1
py36 runtests: PYTHONHASHSEED='3410238989'
py36 runtests: commands[0] | /Development/rust/rust/src/bootstrap/.tox/py36/bin/python bootstrap_test.py
bin_root (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bin_root ... ok
bootstrap_binary (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.bootstrap_binary ... ok
cargo_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.cargo_stamp ... ok
get_mk (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_mk ... ok
get_string (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_string ... ok
get_toml (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.get_toml ... ok
rustc_stamp (bootstrap.RustBuild)
Doctest: bootstrap.RustBuild.rustc_stamp ... ok
format_build_time (bootstrap)
Doctest: bootstrap.format_build_time ... ok
test_stage0_data (__main__.Stage0DataTestCase)
Extract data from stage0.txt ... ok
test_invalid_file (__main__.VerifyTestCase)
Should verify that the file is invalid ... invalid checksum:
    found:    334d016f755cd6dc58c53a86e183882f8ec14f52fb05345887c8a5edd42c87b7
    expected: 64ec88ca00b268e5ba1a35678a1b5316d212f4f366b2477232534a8aeca37f3c
ok
test_valid_file (__main__.VerifyTestCase)
Check if the sha256 sum of the given file is valid ... ok

----------------------------------------------------------------------
Ran 11 tests in 0.019s

OK
___________________________________ summary ____________________________________
  py27: commands succeeded
  py35: commands succeeded
  py36: commands succeeded
  congratulations :)
```

**Comment by aidanhs** [comment]
@milmazz thanks for the PR! We'll make sure @alexcrichton or another reviewer gets to this soon.

One question - as far as I remember we try and keep this compatible with python 3, can you clarify what was broken?

**Comment by milmazz** [comment]
@aidanhs For example:

```console
C:\Development\rust\src\bootstrap>python -V
Python 3.6.1

C:\Development\rust\src\bootstrap>python bootstrap_test.py
Traceback (most recent call last):
  File "bootstrap_test.py", line 21, in <module>
    import bootstrap
  File "C:\Users\tusc0958\Downloads\Development\rust\src\bootstrap\bootstrap.py", line 273
    except OSError, reason:
                  ^
SyntaxError: invalid syntax
```

Also, the `README.md` states that you need `python 2.7` (but not 3.x) in the *Building from Source* section. Should I need to update that part? wdyt?

**Review by alexcrichton — COMMENTED** [review]
Thanks! This part of the codebase very rarely regresses so the need to add tests isn't *too* high, but if you're willing go to through the hoops to add them then I don't see why we shouldn't land them. Can you ensure though that they run as part of our CI?

Otherwise python 3 compatibility is fine, but unfortuantely (last we checked) building LLVM required 2.7 (doesn't work with 3). That being said we can always be compatible for when they're compatible and then we can also provide nicer warnings when 3 is detected.

**Comment by aidanhs** [comment]
I think this PR should also mean that users with a system version of LLVM to link against will be able to build without python 2 installed, which is something (not that system versions of LLVM are recommended).

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -149,25 +150,43 @@ def run(args, verbose=False, exception=False, **kwargs):
...
-        for line in nightlies:
-            line = line.rstrip()  # Strip newline character, '\n'
-            if line.startswith("#") or line == '':
-                continue
-            a, b = line.split(": ", 1)
-            data[a] = b
-    return data
+        lines = [line.rstrip() for line in nightlies
+                 if not line.startswith("#")]
+        return dict([line.split(": ", 1) for line in lines if line])
```
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: Could you move the `if line` into the filtering above (`if line and not line.startswith("#")`)? Seems like it'd make more sense there.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -217,29 +238,37 @@ def download_stage0(self):
...
         default_encoding = sys.getdefaultencoding()
         try:
             ostype = subprocess.check_output(
                 ['uname', '-s']).strip().decode(default_encoding)
-        except (subprocess.CalledProcessError, WindowsError):
+        except subprocess.CalledProcessError:
             return
+        except OSError as reason:
+            if getattr(reason, 'winerror', None) is not None:
+                return
```
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: Surely this needs to `raise` in an else?

File: src/bootstrap/bootstrap.py [hunk 3]
```diff
@@ -285,45 +314,78 @@ def fix_executable(self, fname):
         try:
             subprocess.check_output(
                 ["patchelf", "--set-interpreter", correct_interpreter, fname])
-        except subprocess.CalledProcessError as e:
-            print("warning: failed to call patchelf: %s" % e)
+        except subprocess.CalledProcessError as reason:
+            print("warning: failed to call patchelf:", reason)
```
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: You removed the `%s` on this print, but not the others - I don't think it matters, but it's probably worth being consistent.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (milmazz) [inline src/bootstrap/bootstrap.py]: @aidanhs The way you propose it will break the code when we try to split an empty line. First, we need to delete all the trailing whitespace (eg: `\t\n\x0b\x0c\r `) with `rstrip()` and then we discard all the empty lines.

File: src/bootstrap/bootstrap.py (hunk 2 above)
Comment (milmazz) [inline src/bootstrap/bootstrap.py]: @aidanhs I tried to keep the same behavior, please note that we only catch `subprocess.CalledProcessError`, and `WindowsError` (which only occurs on Windows and it's a subclass of `OSError`). So, if any `OSError` exception occurs in another OS it will certainly raise an exception as you pointed out.

WDYT? Do we need to handle this case in other OSs?

File: src/bootstrap/bootstrap.py (hunk 3 above)
Comment (milmazz) [inline src/bootstrap/bootstrap.py]: @aidanhs Sure thing, I can change those lines too.

File: src/bootstrap/bootstrap.py [hunk 4]
```diff
@@ -149,25 +150,43 @@ def run(args, verbose=False, exception=False, **kwargs):
...
+        self.config_mk = ''
+        self.config_toml = ''
+        self.printed = False
+        self.rust_root = os.path.abspath(os.path.join(__file__, '../../..'))
+        self.use_locked_deps = ''
+        self.use_vendored_sources = ''
+        self.verbose = False
 
     def download_stage0(self):
+        """Download stage0"""
```
Comment (milmazz) [inline src/bootstrap/bootstrap.py]: @alexcrichton Thanks for the review! I agree with you, I added more information in both methods as you suggested. Please let me know what do you think.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: Ah yes, I see.

File: src/bootstrap/bootstrap.py (hunk 2 above)
Comment (aidanhs) [inline src/bootstrap/bootstrap.py]: I'd follow your current theme and leave any behaviour changes for a separate PR, i.e. restore the behaviour of throwing the exception upwards.

**Comment by milmazz** [comment]
@aidanhs @alexcrichton I think this is ready for another review. Please let me know what do you think.

In the meantime, I'll start learning some Rust :-)

File: appveyor.yml [hunk 5]
```diff
@@ -125,6 +125,9 @@ install:
   - copy C:\Python27\python.exe C:\Python27\python2.7.exe
   - set PATH=C:\Python27;%PATH%
 
+  # Run unit tests for bootstrap.py
+  - python src\bootstrap\bootstrap_test.py
```
Comment (aidanhs) [inline appveyor.yml]: Why is this being run during the install phase?

Comment (milmazz) [inline appveyor.yml]: @aidanhs I just thought it was a good place because the 2 preconditions that I need are already meet at this stage:

* [x] The repository is already cloned
* [x] Python is already configured

But I'm open to suggestions, please let me know where do you think is the right place to move this section.

Comment (aidanhs) [inline appveyor.yml]: Wherever it happens, it should probably be in the same place in both travis and appveyor. My personal preference it to put it in the "test" or "script" or whatever phase just for clarity.

Comment (milmazz) [inline appveyor.yml]: In `.travis.yml` we can run custom commands before the `script` phase with `before_script`, it seems that the equivalent phase in `appveyor.yml` is `before_build`.

File: src/bootstrap/bootstrap.py (hunk 2 above)
Comment (alexcrichton) [inline src/bootstrap/bootstrap.py]: @milmazz do you have thoughts on the re-raise here?

Comment (milmazz) [inline src/bootstrap/bootstrap.py]: @alexcrichton I think is not required to re-raise here because if the following condition is `true`:

```python
if getattr(reason, 'winerror', None) is not None:
```

Then, we can agree that we don't need to modify the interpreter section to fix the dynamic linker in the given executable. This condition only applies on NixOS. wdyt?

Comment (milmazz) [inline src/bootstrap/bootstrap.py]: Actually, after thinking a little bit more, you're right, we should have something like this:

```python
except OSError as reason:
	if getattr(reason, 'winerror', None) is not None:
		return
	raise reason
```

If we receive an `OSError` exception that means that the system cannot find the
`uname` command and that is OK in Windows but not in other Operating Systems.

**Comment by carols10cents** [comment]
Ping @aidanhs, looks like this is back in your court for rereview?

File: appveyor.yml [hunk 6]
```diff
@@ -152,6 +152,10 @@ install:
   # Attempt to debug sccache failures
   - set SCCACHE_ERROR_LOG=%CD%/sccache.log
 
+before_build:
+  # Run unit tests for bootstrap.py
+  - python src\bootstrap\bootstrap_test.py
+
```
Comment (aidanhs) [inline appveyor.yml]: Ok I've had a look at this again and realised I missed something - we run Docker containers for a number of our builds and so this python script should ideally run in those as well.

My suggestion is to remove the additions to the travis and appveyor yml, and instead mimic https://github.com/rust-lang/rust/blob/3447aa7/src/ci/run.sh#L65-L69 - hopefully that will work!

**Comment by aidanhs** [comment]
Other than the comment about moving this testing into `run.sh`, I think that's it from me! Let me know once you've had a chance to change that and we can see if the PR build works.

**Comment by milmazz** [comment]
@aidanhs Any idea why the build is failing now? is it related with the latest change?

**Comment by Mark-Simulacrum** [comment]
Build is actually green, just the mac builders were apparently canceled for some reason.

**Comment by milmazz** [comment]
@Mark-Simulacrum Yes, I saw it and I was curious about that. Anyway, I see that a new job was triggered for those cases. Thanks!

**Comment by milmazz** [comment]
Wow, that was fast! now the build is green :)

@aidanhs @alexcrichton  Are we ready to go?

File: src/ci/run.sh [hunk 7]
```diff
@@ -74,6 +74,12 @@ retry make prepare
 travis_fold end make-prepare
 travis_time_finish
 
+travis_fold start configure
```
Comment (aidanhs) [inline src/ci/run.sh]: Can you change the travis folds to be something like `bootstrap-test`?

Comment (milmazz) [inline src/ci/run.sh]: @aidanhs Good Catch! already fixed! 👍

**Comment by aidanhs** [comment]
Looking good! If you can squash your commits then I'll r+.

**Comment by milmazz** [comment]
@aidanhs Done!

**Comment by milmazz** [comment]
@aidanhs Could you please check why the build is failing? nothing appears in the raw log, I did a `rebase` to avoid the merge conflict with this [line](https://github.com/rust-lang/rust/commit/7e6c9f363501c49d3a1f666d85d41891f50890b8#diff-3e203f016947a52e3f1fad3967027e7eR586) in the `bootstrap.py`  file.

**Comment by aidanhs** [comment]
The failures were because the OSX builds didn't start for some reason. They're no-ops anyway, but I restarted them to get the PR build clean.

@bors r+

**Comment by TimNN** [comment]
Looks like a legitimate failure:

```
[00:01:36] /checkout/src/ci/run.sh: line 79: python: command not found
```

**Comment by aidanhs** [comment]
Ok, instead of invoking bootstrap.py directly we need to add a rule to `src/bootstrap/mk/Makefile.in` to be able to use the `$(BOOTSTRAP)`, which has the correct version of Python detected for us.

I reckon just add a `check-bootstrap` above the `check-aux` rule - https://github.com/rust-lang/rust/blob/04415dc/src/bootstrap/mk/Makefile.in#L55

**Comment by Mark-Simulacrum** [comment]
`[00:02:06] /checkout/src/ci/run.sh: line 79: python: command not found` -- I suspect something isn't using the configure script's output to find python...

File: src/ci/run.sh [hunk 8]
```diff
@@ -74,6 +74,12 @@ retry make prepare
 travis_fold end make-prepare
 travis_time_finish
 
+travis_fold start bootstrap-test
+travis_time_start
+python $SRC/src/bootstrap/bootstrap_test.py
```
Comment (Mark-Simulacrum) [inline src/ci/run.sh]: Let's add this as a target to the makefile. I think you can edit `Makefile.in` to do so, and then we can use the `CFG_PYTHON` and `CFG_SRC_DIR` variables. `check-bootstrap` seems fine.

**Comment by aidanhs** [comment]
Wait, I don't understand why bors tested this again? I requested a change to put this in the makefile three days ago, this PR was assigned to me two days ago...and then bors tested it an hour ago?

**Comment by Mark-Simulacrum** [comment]
Huh. Possibly because we synchronized bors' queue and it somehow forgot about the failure, but odd indeed.

**Comment by milmazz** [comment]
@aidanhs @Mark-Simulacrum Done! 🤞

**Comment by aidanhs** [comment]
Thanks!

@bors r+

**Comment by aidanhs** [comment]
@bors retry

 - #43453

**Comment by Mark-Simulacrum** [comment]
@bors retry
 - RLS spurious failure - #43453

```
01:32:40] test test::test_borrow_error ... FAILED
[01:32:40] 
[01:32:40] failures:
[01:32:40] 
[01:32:40] ---- test::test_borrow_error stdout ----
[01:32:40] 	thread 'test::test_borrow_error' panicked at 'Hit timeout', /checkout/src/tools/rls/src/test/harness.rs:142:12
[01:32:40] 
```

**Comment by milmazz** [comment]
@aidanhs @alexcrichton @Mark-Simulacrum Thank you for your help! :tada: :man_dancing: :tada:
